guac_core/tests/contract_test.rs

```diff
@@ -0,0 +1,470 @@
+extern crate clarity;
+extern crate guac_core;
+extern crate web3;
+#[macro_use]
+extern crate lazy_static;
+extern crate rand;
+#[macro_use]
+extern crate failure;
+
+use clarity::abi::{derive_signature, encode_call, encode_tokens, Token};
+use clarity::{Address, BigEndianInt, PrivateKey, Transaction};
+use failure::Error;
+use guac_core::channel_client::channel_manager::ChannelManager;
+use guac_core::crypto::CryptoService;
+use guac_core::crypto::CRYPTO;
+use guac_core::eth_client::create_open_channel_payload;
+use rand::{OsRng, Rng};
+use std::env;
+use std::ops::Deref;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::time;
+use web3::futures::future::ok;
+use web3::futures::Async;
+use web3::futures::{Future, IntoFuture, Stream};
+use web3::transports::{EventLoopHandle, Http};
+use web3::types::{Bytes, FilterBuilder, Log, TransactionRequest, H160, U256};
+use web3::Web3;
+
+/// A handle that contains event loop instance and a web3 instance
+///
+/// EventLoop has to live at least as long as the "Web3" object, or
+/// otherwise calls will fail. We achieve this by implementing a Deref
+/// trait that would return a borrowed Web3 object.
+struct Web3Handle(EventLoopHandle, Web3<Http>);
```

Comment:
If it needs to stay though the Deref trait obfuscates the fact that all methods are called on `Web3`

Reply:
Its purpose is to contain EventLoopHandle and Web3<Http> together so they have the same lifetime. Without alive event loop any web3 call would fail. So it is in fact hiding the existence of a EventLoopHandle, and the Web3Handle is exposed through a global variable `WEB3`.